### ISSUE TITLE
Throwing exceptions across extension host RPC boundaries doesn't preserve Error.name

### ISSUE THREAD

**Issue by AddieCohen** [microsoft/vscode#156281]
Does this issue occur when all extensions are disabled?: No

- VS Code Version: 1.69.2
- OS Version: macOS 12.4 arm64

Steps to Reproduce:

1. Clone https://github.com/AddieCohen/auth-provider-error 
2. `npm install` & run the extension.
3. Run the command "Login with Azure DevOps"
4. Check the debug console, the error name property is "Error" instead of `LoginTimeOutError`.

The line in: https://github.com/microsoft/vscode/blob/2980862817f29911a2231f4c88bfc783bd763cab/src/vs/workbench/services/extensions/common/rpcProtocol.ts#L428 doesn't seem to work to preserve the exception name.

**Comment by alexdima** [comment]
Unfortunately we cannot serialize/deserialize arbitrary `Error` subclasses across RPC boundaries. Having a certain subclass of `Error` does not automatically lead to the `.name` property to be set to the subclass name. e.g.

<img width="268" alt="image" src="https://user-images.githubusercontent.com/5047891/181255749-ffc3a1bc-a761-40cf-97cb-1bc7c1e0a4c7.png">

**Comment by segevfiner** [comment]
There is a line in the RPC code that tries to preserve the `name` property but it isn't working, see the issue description.

**Comment by alexdima** [comment]
@segevfiner Please see the screenshot I posted and note how the `name` property is always `Error` even for a custom error class that subclasses `Error`

**Comment by segevfiner** [comment]
Exactly. See the issue description, this line tries to preserve it: https://github.com/microsoft/vscode/blob/2980862817f29911a2231f4c88bfc783bd763cab/src/vs/workbench/services/extensions/common/rpcProtocol.ts#L428

But it doesn't work. Probably an `Object.defineProperty` blocking changing it?

**Comment by alexdima** [comment]
@segevfiner The OP is doing debugging and is under the impression that the debugger displays the `.name` property of an error in the debugger. But that is false, the debugger displays the ctor name.

The following code snippet demonstrates that simply subclassing from `Error` does not define the `.name` property:
```ts
export class LoginTimeOutError extends Error {
    constructor(message?: string) {
        super(message);
    }
}
const y = new LoginTimeOutError();
console.log(y.name); // prints Error
```

In order for the `.name` property to be changed, a manual step must be taken by whomever is writing the custom class e.g.:
```ts
export class LoginTimeOutError extends Error {
    constructor(message?: string) {
        super(message);
        this.name = 'LoginTimeOutError';
    }
}
const y = new LoginTimeOutError();
console.log(y.name); // prints LoginTimeOutError
```

So the problem here is that the OP and you believe that there is a problem with the RPC code, but the RPC code is doing its job just fine, it serializes the name `Error` and restores the name `Error`. The root problem is that the initial error doesn't have a custom name.

**Comment by segevfiner** [comment]
This could be it. Do we have any docs about this? If we don't, can we improve the documentation so it's obvious that this how this should be used?

**Comment by alexdima** [comment]
PR to improve docs are always welcome. @jrieken do you have any ideas where we could document how errors are transported across processes?
